### PR TITLE
Encoding Subtype property as NSString rather than NS_ENUM

### PR DESCRIPTION
**Pull request by fleupold**
As discussed with @andrewcuneo, using NS_ENUM to encode the subtype of ADTs is a problem if we ever change the definition of .adtValue file and try to read persisted instances of the older version.

Therefore, I changed the NSCoding plugin to encode the subtype as a String rather than an enum value. I am not sure what we should do if we try to deserialize an old type - return nil vs throw or something else. Happy to discuss.

Unit and acceptance tests are adjusted and passing.

### CODE REVIEW

File: src/plugins/coding.ts [hunk 1]
```diff
@@ -381,6 +405,28 @@ function algebraicAttributeToUnsupportedTypeError(algebraicType:AlgebraicType.Ty
...
+  return constantValueForAttributeName('SUBTYPE_' + AlgebraicTypeUtils.subtypeNameFromSubtype(subtype));
+}
+
+function codeForSubtypeBranchesWithSubtypeMapper(algebraicType:AlgebraicType.Type, subtypeValueAccessor:string, subtypeMapper:(algebraicType:AlgebraicType.Type, subtype:AlgebraicType.Subtype) => string[], soFar:string[], subtype:AlgebraicType.Subtype):string[] {
+  const internalCode:string[] = subtypeMapper(algebraicType, subtype);
+  const code:string[] = [(soFar.length ? 'else if([' : 'if([') + subtypeValueAccessor + ' isEqualToString:' + CodingNameForSubtype(subtype) + ']) {'].concat(internalCode.map(StringUtils.indent(2))).concat(['}']);
+  return soFar.concat(code);
+}
+
+function p3Applyf5<T,U,V,W,X,Y>(firstVal:T, secondVal:U, thirdValue:V, f:(a:T, b:U, c:V, d:W, e:X) => Y):(d:W, e:X) => Y {
```
Comment (andrewcuneo) [inline src/plugins/coding.ts]: Is this file the right place for this?

File: features/algebraic-types-coding.feature [hunk 2]
```diff
@@ -83,17 +84,18 @@ Feature: Outputting Algebraic Types
...
-            case _SimpleADTSubtypesFirstSubtype: {
-              _firstSubtype_firstValue = [aDecoder decodeObjectForKey:kFirstSubtypeFirstValueKey];
-              _firstSubtype_secondValue = [aDecoder decodeIntegerForKey:kFirstSubtypeSecondValueKey];
-              break;
-            }
-            case _SimpleADTSubtypesSecondSubtype: {
-              _secondSubtype_something = [aDecoder decodeBoolForKey:kSecondSubtypeSomethingKey];
-              break;
-            }
+          _codedSubtype = [aDecoder decodeObjectForKey:kCodedSubtypeKey];
```
Comment (andrewcuneo) [inline features/algebraic-types-coding.feature]: Can't this be an method variable instead of a class variable?

**Comment by andrewcuneo** [comment]
Overall, looks really good. A couple of questions.

File: src/plugins/coding.ts (hunk 1 above)
Comment (fleupold) [inline src/plugins/coding.ts]: Moving to function-util.ts

File: features/algebraic-types-coding.feature (hunk 2 above)
Comment (fleupold) [inline features/algebraic-types-coding.feature]: It can. I tried to reuse as much of the existing code as possible, but will change.

**Pull request merged**